File: web/tests/go/controller/session_stats_monotonic_test.go
package controller_test

import (
	"encoding/json"
	"testing"
)

// AUX-15: an accumulated usage snapshot that shrinks at compaction must not
// regress the totals, while the live context usage still tracks the SDK.
func TestUsageTotalsStayMonotonicAcrossCompaction(t *testing.T) {
	loadUpdates := []map[string]any{
		{"__piOnly": true, "sessionUpdate": "usage_update",
			"accumulatedInputTokens": 1000, "accumulatedOutputTokens": 500,
			"accumulatedCost": 1.5, "costCurrency": "USD",
			"contextLimit": 200000, "used": 1800},
	}
	manager, _, project, _ := newSessionManager(t, loadUpdates, nil)
	if _, err := manager.Messages(t.Context(), "chat", project.ID, project.Roots[0], "test"); err != nil {
		t.Fatal(err)
	}
	before, err := manager.Stats("chat")
	if err != nil {
		t.Fatal(err)
	}
	if before.Tokens.Input != 1000 || before.Tokens.Output != 500 || before.Cost != 1.5 {
		t.Fatalf("initial usage projection: %#v", before)
	}

	// Compaction re-emits a smaller accumulated snapshot.
	payload, err := json.Marshal(map[string]any{
		"sessionId": "chat",
		"update": map[string]any{
			"sessionUpdate":           "usage_update",
			"accumulatedInputTokens":  400,
			"accumulatedOutputTokens": 100,
			"accumulatedCost":         0.5,
			"contextLimit":            200000,
			"used":                    800,
		},
	})
	if err != nil {
		t.Fatal(err)
	}
	if err := manager.Extension(t.Context(), "pi.session.update", payload); err != nil {
		t.Fatal(err)
	}
	after, err := manager.Stats("chat")
	if err != nil {
		t.Fatal(err)
	}
	if after.Tokens.Input != 1000 || after.Tokens.Output != 500 || after.Cost != 1.5 {
		t.Fatalf("compaction regressed the totals: %#v", after)
	}
	if tokens, _ := after.ContextUsage["tokens"].(int64); tokens != 800 {
		t.Fatalf("live context usage did not follow the SDK: %#v", after.ContextUsage)
	}
}

File: tests/webui/chat/session-stats-monotonic.test.ts
import { expect, test } from "bun:test";
import {
	createSessionRuntime,
	reduceSessionEvent,
} from "../../../webui/src/chat/runtime/session-runtime";
import { mergeUsageStats } from "../../../webui/src/chat/session/session-stats";

test("compaction usage does not regress totals while context follows the SDK", () => {
	let runtime = createSessionRuntime(null, "off");
	runtime = reduceSessionEvent(runtime, {
		type: "usage",
		usage: { input: 1_000, output: 500, cacheRead: 200, cacheWrite: 100, total: 1_800, cost: 1.5 },
		costCurrency: "USD",
		reported: { input: true, output: true, total: true, cost: true },
	});
	runtime = reduceSessionEvent(runtime, {
		type: "context",
		contextUsage: { tokens: 1_800, contextWindow: 200_000, percent: 0.9 },
	});
	// A compaction removes older messages and re-emits a smaller accumulated
	// snapshot. Every prior entry still contributed, so the totals hold.
	runtime = reduceSessionEvent(runtime, {
		type: "usage",
		usage: { input: 400, output: 100, cacheRead: 200, cacheWrite: 100, total: 800, cost: 0.5 },
		reported: { input: true },
	});
	expect(runtime.stats?.tokens.input).toBe(1_000);
	expect(runtime.stats?.tokens.output).toBe(500);
	expect(runtime.stats?.tokens.cacheRead).toBe(200);
	expect(runtime.stats?.tokens.cacheWrite).toBe(100);
	expect(runtime.stats?.tokens.total).toBe(1_800);
	expect(runtime.stats?.cost).toBe(1.5);
	expect(runtime.stats?.contextUsage).toEqual({
		tokens: 1_800,
		contextWindow: 200_000,
		percent: 0.9,
	});
	// Live context usage is the SDK value and may legitimately shrink.
	runtime = reduceSessionEvent(runtime, {
		type: "context",
		contextUsage: { tokens: 800, contextWindow: 200_000, percent: 0.4 },
	});
	expect(runtime.stats?.contextUsage).toEqual({
		tokens: 800,
		contextWindow: 200_000,
		percent: 0.4,
	});
	expect(runtime.stats?.tokens.input).toBe(1_000);
});

test("mergeUsageStats keeps context usage out of the merge", () => {
	const merged = mergeUsageStats(
		{
			sessionId: "s",
			totalMessages: 3,
			tokens: { input: 10, output: 5, cacheRead: 1, cacheWrite: 1, total: 17 },
			cost: 0.25,
			costCurrency: "USD",
			contextUsage: { tokens: 42, contextWindow: 100, percent: 42 },
		},
		{ usage: { input: 4, output: 2, total: 6, cost: 0.1 } },
	);
	expect(merged.tokens).toEqual({ input: 10, output: 5, cacheRead: 1, cacheWrite: 1, total: 17 });
	expect(merged.cost).toBe(0.25);
	expect(merged.costCurrency).toBe("USD");
	expect(merged.contextUsage).toEqual({ tokens: 42, contextWindow: 100, percent: 42 });
});
